### PR TITLE
Revert "Samplers: don't create unneeded empty samplers during startup"

### PR DESCRIPTION
**Pull request by ronso0**
This reverts the core commit of #12657 

Besides the Numark Mixtrack Pro3 #12808 other mappings slipped through in #12769.
I'm at it but I can't guarantuee to fix _all_ of them _in time_.
Also, user mappings may be broken, too.

Let's push #12657 to 2.4.1

To summarize: the mapping control error can still occur if
* _samplers.xml_ doesn't exist / is broken
* loading a mapping that needs N samplers
* before loading a skin that provides N samplers

### CODE REVIEW

**Comment by ronso0** [comment]
I think in order to not require users to create expected decks and samplers explicitly in script files this could be done where control existence is checked in `ControllerScriptInterfaceLegacy`:
* if control refers to a deck/sampler > current deck/sampler count
   * if control exists for deck/sampler **1**
      * create deck/sampler
   * else throw error
* else throw error

**Comment by daschuer** [comment]
Oh, I am sorry this topic has put extra workload stress on you.

I am not sure if merging this is a good idea. Do we hunt one regression (now only in custom mappings) with another, the performance regression and the reason we have created dedicated sampler skins? 

What is the effect when a mapping requests missing samplers? I think using a sampler without a GUI is unlikely. The assertion about the missing control is easy to fix. Probably more easy than a latency increase. 

After merging this, we have hard time to identify the not updated mappings. 

In general it is probably also questionable if mappings should be allowed to control samplers without a GUI.

**Comment by ronso0** [comment]
> Oh, I am sorry this topic has put extra workload stress on you.

Never mind, it was my mistake to not do a smoke test with my regular setup. I'd have noticed right away.

Tbh I didn't measure the performance implication of 64 unneeded samplers. Should be about more memory being used, not really CPU load (no sampler loaded --> no caching reader work, no engine buffer work IIUC).
And no one noticed or cared about for years, so I think this is no big deal for now.

> What is the effect when a mapping requests missing samplers? I think using a sampler without a GUI is unlikely. 

The GUI samplers are created _after_ the mapping has been loaded, that's the issue.
That's the sample deck reation order:
* `CoreServices::initialize` triggers creation of the default samplers (4 currently) [here, line 309](https://github.com/mixxxdj/mixxx/blob/170987331434fc33becaabe13d6c0446bc107023/src/coreservices.cpp#L309-L311)
* a bit later, same function, _samplers.xml_ is read [here, line 431](https://github.com/mixxxdj/mixxx/blob/170987331434fc33becaabe13d6c0446bc107023/src/coreservices.cpp#L431)
   for example: 64 samplers, tracks in samplers 1-12, = 12 samplers are created
* (MixxxMainWindow is initialized)
* main.cpp triggers controller setup [here](https://github.com/mixxxdj/mixxx/blob/170987331434fc33becaabe13d6c0446bc107023/src/main.cpp#L98), mappings are loaded 
* skin is loaded, skin manifest requests 16 samplers in [`LegacySkinParser::parseSkin`](https://github.com/mixxxdj/mixxx/blob/170987331434fc33becaabe13d6c0446bc107023/src/skin/legacy/legacyskinparser.cpp#L346)


> The assertion about the missing control is easy to fix.

Not sure what you have in mind. Something like I expressed above  https://github.com/mixxxdj/mixxx/pull/12809#issuecomment-1941111344

> In general it is probably also questionable if mappings should be allowed to control samplers without a GUI.

Hmm, not sure about this.
Also, that can not be detected reliably: we know if a mapping conatins a direct xml mapping for a deck/sampler control, or tries to make a connection to one, but such a control  also be in a mapping function and IIUC we would only know about this when the function is executed because that's when it's evaluated.

**Bottom line:**
I recommend to merge this and work on a fix for 2.4.1 or 2.5

**Review by daschuer — APPROVED** [review]
OK, than lets merge it. Thank you for explaining.
